plugin/pulpcore/plugin/download/asyncio/group.py

```diff
@@ -0,0 +1,190 @@
+import asyncio
+from collections import defaultdict
+from contextlib import suppress
+from itertools import chain
+
+from .http import HttpDownloader
+from .factory import DownloadFactory
+
+
+class GroupDownloader(object):
+    """
+    Download groups of files, with all downloads from all groups occurring in parallel.
+
+    This supports groups to be downloaded by adding them explicitly or via a iterator such as a
+    generator. The generator case is useful for limiting the number of in-memory objects.
+
+    Basic Usage:
+        >>> downloader = GroupDownloader(importer)
+        >>> artifact_a = DeferredArtifact(url=url_a md5='912ec803b2ce49e4a541068d495ab570')
+        >>> artifact_b = DeferredArtifact(url=url_b, size=4172)
+        >>> downloader.schedule_group('my_id', [artifact_a, artifact_b])
+        >>> for id, group_results in downloader:
+        >>>     print(id)  # id is set to 'my_id'
+        >>>     # group_results is a dict of DownloadResult objects keyed by DeferredArtifact
+
+    If you register a large number of groups, you could use a lot of memory holding those objects,
+    their associated downloaders, and other related objects. To resolve this, the GroupDownloader
+    provides a way to limit the number of effective in-memory objects by processing objects from a
+    generator that yields the (id, [DeferredArtifact()]) tuples. See the example below and read the
+    schedule_from_iterator() docs for more details.
+
+    Constraining in-memory objects while downloading:
+        >>> def group_generator():
+        >>>    yield (id, [DeferredArtifact(url=url_a), DeferredArtifact(url=url_b)])
+        >>>    yield (id, [DeferredArtifact(url=url_a), DeferredArtifact(url=url_b)])
+        >>>
+        >>> downloader = GroupDownloader(importer)
+        >>> downloader.schedule_from_iterator(group_generator)
+        >>> for id, group_results in downloader:
+        >>>     print(id)  # id is set to 'my_id'
+        >>>     # group_results is a dict of DownloadResult objects keyed by DeferredArtifact
+    """
+
+    def __init__(self, importer, download_class=HttpDownloader):
+        """
+        Args:
+            importer (pulpcore.plugin.models.Importer): The importer used to configure downloaders
+                with.
+            download_class (pulpcore.plugin.download.HttpDownloader): The asyncio-aware downloader
+                class the GroupDownloader will use when instantiating downloaders. This option
+                allows the GroupDownloader to work with custom downloaders.
+        """
+        self.importer = importer
+
+        self.download_factory = DownloadFactory(self.importer, download_class=download_class)
```

Comment:
Suggest using `Importer.get_asyncio_downloader()` instead using `DownloadFactory` and `download_class`